src/Microsoft.Diagnostics.Monitoring/DiagnosticsEventPipeProcessor.cs

```diff
@@ -470,17 +470,36 @@ private async Task HandleGCEvents(EventPipeEventSource source, int pid, Func<Tas
             _gcGraph.AllowReading();
         }
 
-        private void HandleProcessInfo(EventPipeEventSource source, Func<Task> stopFunc, CancellationToken token)
+        private async Task HandleProcessInfo(EventPipeEventSource source, Func<Task> stopFunc, CancellationToken token)
         {
-            source.Dynamic.AddCallbackForProviderEvent(MonitoringSourceConfiguration.EventPipeProviderName, "ProcessInfo", traceEvent =>
+            Action<TraceEvent, Action> processInfoHandler = (TraceEvent traceEvent, Action taskComplete) =>
             {
                 _processInfoCallback?.Invoke((string)traceEvent.PayloadByName("CommandLine"));
-            });
-
-            source.Dynamic.All += traceEvent =>
-            {
-                stopFunc();
+                taskComplete();
             };
+
+            // Completed when the ProcessInfo event of the Microsoft-DotNETCore-EventPipe event provider is handled
+            using var processInfoTaskSource = new EventTaskSource<Action<TraceEvent>>(
+                taskComplete => traceEvent => processInfoHandler(traceEvent, taskComplete),
+                handler => source.Dynamic.AddCallbackForProviderEvent(MonitoringSourceConfiguration.EventPipeProviderName, "ProcessInfo", handler),
+                handler => source.Dynamic.RemoveCallback(handler),
+                token);
+
+            // Completed when any trace event is handled
```

Comment:
why do we have this?

Reply:
The ProcessInfo event is not provided to the event stream until the event pipe session is requested to be shut down. Thus, we need to start the session, wait for any event to be provided, and then shut down the session, in order to get the ProcessInfo event to be produced.